### PR TITLE
Add GitHub Oauth Support

### PR DESCRIPTION
**Pull request by drasticactions**
This update adds GitHub OAuth Support to Wrench. By adding the needed Client/Secret IDs to the MonkeyWrench config file, you can allow users to use their GitHub credentials to log into Wrench. With the bonus of now having their GitHub access token, which we could use in Wrench.

Like with OpenID/Google Auth, you can add which team organizations a user is in to give them appropriate credentials.

To add support for this, I used a [forked library] (https://github.com/drasticactions/GitHubOAuth2Client) that adds GitHub support to DotNetOpenAuth. It's a simple class library that's not available on Nuget, and I updated it to provide scope support and let us access team organizations easier. The base library is MIT licensed.

### CODE REVIEW

File: MonkeyWrench/Configuration.cs [hunk 1]
```diff
@@ -241,7 +248,7 @@ public static bool LoadConfiguration (string [] arguments, string file)
 				UploadPort = int.Parse (xml.SelectSingleNode ("MonkeyWrench/Configuration/UploadPort").GetNodeValue (UploadPort.ToString ()));
 				AllowAnonymousAccess = bool.Parse(xml.SelectSingleNode("MonkeyWrench/Configuration/AllowAnonymousAccess").GetNodeValue(AllowAnonymousAccess.ToString()));
 				OpenIdProvider = xml.SelectSingleNode ("MonkeyWrench/Configuration/OpenIdProvider").GetNodeValue (OpenIdProvider);
-				OpenIdRoles = xml.SelectSingleNode ("MonkeyWrench/Configuration/OpenIdRoles").GetNodeValue (OpenIdRoles);
+				OpenIdRoles = xml.SelectSingleNode ("MonkeyWrench/Configuration/OpenIdRoles").GetNodeValue ("OpenIdRoles").Split(';');
```
Comment (rolfbjarne) [inline MonkeyWrench/Configuration.cs]: There might not be a `OpenIdRoles` node, in which case this will cause a NullReferenceException.

Comment (drasticactions) [inline MonkeyWrench/Configuration.cs]: Good point, I'll fix it and GitHubOrganizationList, since they both do the same thing.

File: MonkeyWrench.Database/Extensions/DBLogin_Extensions.cs [hunk 2]
```diff
@@ -69,33 +71,41 @@ public static DBLogin Login (DB db, string login, string password, string ip4, b
...
-		public static void LoginOpenId (DB db, LoginResponse response, string email, string ip4)
+		public static void Login (DB db, LoginResponse response, string email, string ip4, List<string> userOrgs, bool useGitHub = false, string gitHubLogin = "")
 		{
-			if (string.IsNullOrEmpty (Configuration.OpenIdProvider) && string.IsNullOrEmpty (Configuration.OauthClientId))
-				throw new Exception ("No OpenId provider available");
+			string [] specs;
 
-			if (string.IsNullOrEmpty (Configuration.OpenIdRoles))
-				throw new Exception ("No OpenId roles specified");
+			string username = useGitHub ? gitHubLogin : email;
```
Comment (rolfbjarne) [inline MonkeyWrench.Database/Extensions/DBLogin_Extensions.cs]: There's a possible attack vector here, if someone manages to create a github account that matches an openid email.

That's currently not possible (`@` isn't allowed in github user names), but there's also no guarantees about the future.

I think the simplest solution for now would be to validate that a github login does not contain a `@`, and at the same time validate that an email does contain a `@`. That way openid and github logins would fall into two disjoint domains.

File: MonkeyWrench/MonkeyWrench.csproj [hunk 3]
```diff
@@ -62,7 +62,6 @@
     </Reference>
   </ItemGroup>
   <ItemGroup>
-    <Compile Include="AssemblyInfo.cs" />
```
Comment (rolfbjarne) [inline MonkeyWrench/MonkeyWrench.csproj]: Was this intentional?

Comment (drasticactions) [inline MonkeyWrench/MonkeyWrench.csproj]: This was intentional because it broke my build otherwise. There was no AssemblyInfo file in the solution, I had to remove it in order to build at all. This _could_ be moved into a separate commit though, or was the file meant to be in the repo?

Comment (rolfbjarne) [inline MonkeyWrench/MonkeyWrench.csproj]: It's generated: https://github.com/mono/monkeywrench/blob/master/MonkeyWrench/Makefile#L2

Comment (drasticactions) [inline MonkeyWrench/MonkeyWrench.csproj]: Fixed in https://github.com/mono/monkeywrench/pull/74/commits/7121a27d784dc681fae54903a1ae59c42eb73964

File: MonkeyWrench.Database/Extensions/DBLogin_Extensions.cs (hunk 2 above)
Comment (drasticactions) [inline MonkeyWrench.Database/Extensions/DBLogin_Extensions.cs]: Fixed in https://github.com/mono/monkeywrench/pull/74/commits/7121a27d784dc681fae54903a1ae59c42eb73964

Comment (drasticactions) [inline MonkeyWrench.Database/Extensions/DBLogin_Extensions.cs]: Actually I thought about this some more and talked with @duncanmak about it to make sure this filter made sense. I don't think it does, because `username` is not used as a way to authenticate. It's only used as a way to give the user a username. The GitHub check is based on their organization, so they have to be in a organization that is listed in the Wrench config file. Email is used if you are using Google/OpenID. And you can only use email if the `useGitHub` flag is set to false.

I don't think there is an attach issue that would happen here. And while the filter is harmless, it can be confusing because there is no need for it and can make people looking at the code think that `username` means something it's not.

I removed the filter here [https://github.com/mono/monkeywrench/pull/74/commits/cd6acb7ede2f142d723ed0f28c6ae4ee1e708e8c] .

**Comment by drasticactions** [comment]
@rolfbjarne So we should be good to look at this again if you have the chance.

**Comment by rolfbjarne** [comment]
Sorry, I thought I had merged it already. Now it's done.
